File: tests/pendulum/py_pendulum.py
import unittest
from collections import OrderedDict

from libpy_pendulum import InvertedPendulum, SignalBase

MEMBERS = OrderedDict([('cart_mass', 4.2), ('pendulum_mass', 2.3), ('pendulum_length', 0.8), ('viscosity', 0.5)])


class TestInvertedPendulum(unittest.TestCase):
    def test_inverted_pendulum(self):
        name = 'test inverted_pendulum'
        ip = InvertedPendulum(name)

        self.assertEqual(ip.name, name)
        self.assertEqual(ip.class_name, 'InvertedPendulum')
        self.assertEqual(ip.docstring, 'Classical inverted pendulum dynamic model\n')

        for member, value in MEMBERS.items():
            setattr(ip, member, value)
            self.assertEqual(getattr(ip, member), value)

        for _ in range(100):
            ip.incr(1e-3)

        self.assertTrue(ip.has_signal('state'))
        self.assertTrue(ip.has_signal('force'))
        self.assertFalse(ip.has_signal('plop'))

        ip2 = InvertedPendulum(name + '2', *MEMBERS.values())
        for member, value in MEMBERS.items():
            self.assertEqual(getattr(ip2, member), value)

        force_in = ip.force
        self.assertTrue(isinstance(force_in, SignalBase))
        self.assertEqual(force_in.time, 0)

        # state_out = ip.state
        # self.assertTrue(isistance(state_out, SignalBase))


if __name__ == '__main__':
    unittest.main()

File: tests/pendulum/py_pendulum.cpp
#include <boost/python.hpp>
#include <py-dynamic-graph/py-dynamic-graph.hpp>

#include "pendulum.hpp"

BOOST_PYTHON_MODULE(libpy_pendulum)
{
  using namespace boost::python;
  using namespace dynamicgraph;

  expose_signal_base();
  expose_entity();
  register_types();

  class_<InvertedPendulum, bases<Entity>, boost::noncopyable>("InvertedPendulum", init<std::string, optional<double, double, double, double> >())
    .add_property("cart_mass", &InvertedPendulum::getCartMass, &InvertedPendulum::setCartMass)
    .add_property("pendulum_mass", &InvertedPendulum::getPendulumMass, &InvertedPendulum::setPendulumMass)
    .add_property("pendulum_length", &InvertedPendulum::getPendulumLength, &InvertedPendulum::setPendulumLength)
    .add_property("viscosity", &InvertedPendulum::getViscosity, &InvertedPendulum::setViscosity)
    .def("incr", &InvertedPendulum::incr)
    .add_property("force", &InvertedPendulum::getForce )
    .add_property("state", &InvertedPendulum::getState )
    ;
}
